main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
+          evalLog.debug(s"Scheduling ${newInProgress.size} new tasks (issuer: ${issuer}): ${newInProgress.map(t => printTerm(t._1)).mkString(", ")}")
+
+          // schedule for parallel execution
+          newInProgress.zipWithIndex.foreach {
+            case (curWork @ (terminal, group), index) =>
+
+              //              val missingDependencies = group.indexed.flatMap(_.inputs).filter(t => !results.contains(t) && !group.contains(t))
+              //              if (!missingDependencies.isEmpty) {
+              //                evalLog.error(s"Missing resolved dependencies for terminal group: ${printTerm(terminal)}\n  ${missingDependencies.mkString(",\n  ")}")
+              //              }
+
+              val workerFut: java.util.concurrent.Future[FutureResult] = completionService.submit { () =>
+                if(failFast && someTaskFailed) {
+                  // we do not start this tasks but instead return with aborted result
+                  val newResults = group.map { task =>
+                    task -> Aborted
+                  }
+
+                  evalLog.debug(s"Skipped evaluation (because of earlier failures): ${printTerm(terminal)}")
+                  FutureResult(curWork, 0, Evaluated(newResults.toMap, Seq(), false))
+
+                } else {
+
+                  val counterMsg = nextCounterMsg()
+
+                  evalLog.debug(s"Start evaluation [${counterMsg}]: ${printTerm(terminal)}")
+                  val startTime = System.currentTimeMillis()
+
+                  val res = evaluateGroupCached(
+                    terminal,
+                    group,
+                    results,
+                    counterMsg,
+                    reporter,
+                    testReporter,
+                    logger
+                  )
+
+                  val endTime = System.currentTimeMillis()
+                  evalLog.debug(s"Finished evaluation [${counterMsg}]: ${printTerm(terminal)}")
+
+                  FutureResult(curWork, (endTime - startTime).toInt, res)
+                }
+              }
+
+              evalLog.debug(s"New future: ${workerFut} [${index + 1}/${newInProgress.size}] for task: ${printTerm(terminal)}")
+              futures = futures ++ List(workerFut -> curWork)
+          }
+        } else {
+          evalLog.debug(s"No new tasks to schedule (issuer: ${issuer})")
+        }
+      }
+
+      scheduleWork("initial request")
+
+      // Work queue management
+      // wait for finished jobs and schedule more work, if possible
+      while (futures.size > 0) {
+        evalLog.debug(s"Waiting for next future completion of ${executorService}")
+        val compFuture: Future[FutureResult] = completionService.take()
+
+        val compTask = futures.find(_._1 == compFuture).map(_._2).get
+        val compTaskName = printTerm(compTask._1)
+        evalLog.debug(s"Completed future: ${compFuture} for task ${compTaskName}")
+        futures = futures.filterNot(_._1 == compFuture)
```

Comment:
It looks like we're basically using this `futures: List[_]` as a queue; should we use a `mutable.Queue` instead?

Reply:
Yes